### PR TITLE
SearchKit - Add API filter for contacts in groups and smart groups

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Adds a filter-style field to search for contacts in (or not in) one or more groups.

Before
----------------------------------------
Not possible to search for contacts in groups.

After
----------------------------------------
Regular groups, smart groups & nested groups all work as SearchKit filters.

image

Technical Details
----------------------------------------
Adds some new concepts in APIv4 getFields:
1. Now a field can have a "type". The types are:
    - **Field:** core field
    - **Custom:** custom field
    - **Filter:** field that can only be used in the Where clause
    - **Extra:** any other ad-hoc api field
2. Fields can now have `sql_filters` which are php callbacks to generate custom sql.
3. Fields can specify which operators they support.

### CODE REVIEW

File: Civi/Api4/Service/Spec/Provider/ContactGetSpecProvider.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    // SQL optimization - use INNER JOIN if the base table is Contact & this clause is not nested
+    if ($fieldAlias === '`a`.`id`' && $operator === "IN" && !$depth) {
+      $query->getQuery()->join($tableName, "INNER JOIN `$tableName` ON $fieldAlias = `$tableName`.contact_id");
+      return '1';
+    }
+    // Else use IN or NOT IN (this filter only supports those 2 operators)
+    else {
+      return "$fieldAlias $operator (SELECT contact_id FROM `$tableName`)";
+    }
+  }
```
Comment (colemanw) [inline Civi/Api4/Service/Spec/Provider/ContactGetSpecProvider.php:76]: @eileenmcnaughton this is the meat of it. The rest is to support using a callback like this to supply sql for a field.

**Comment by seamuslee001** [comment]
Test fails I believe relate here

**Comment by eileenmcnaughton** [comment]
@colemanw is this testable via search kit now? (In which case I should test by trying to create searches on our testing server to give it some performance work to do)

**Comment by colemanw** [comment]
@eileenmcnaughton yes have fun :)

**Comment by eileenmcnaughton** [comment]
Ok - even if this is merged post rc forking (which seems likely based on current test fails) - I'll deploy it with the 5.39 rc to our site so it gets a good kicking

**Comment by eileenmcnaughton** [comment]
This worked in my tests so far -I'm going to keep testing but I think anything else can be as a follow up as I think the approach won't get us in trouble - ie the surface area of this change is limited to adding the group filter to the api contract.

I do think we need a push into documentation next....

**Pull request merged**